### PR TITLE
	modified:   src/v2i-hub/ImmediateForwardPlugin/manifest.json

### PR DESCRIPTION
**Pull request by jwillmartin**
modified:   src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.cpp
	modified:   src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h

# PR Details
## Description

Updated IFM plugin based on feedback from Plugfest testing and other users. There's a need to change the Active Message File contents that the IFM plugin generates. These additions allows users to configure the TxMode and Signature fields without changing other functionality to the plugin.

## Motivation and Context

Users need to be able to configure some fields in the Immediate Forward Active Message File, which were previously hard-coded. 

## How Has This Been Tested?

Built docker image locally and tested that inputs to the new configurations were inserted into the Active Message File as expected. 

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/v2i-hub/ImmediateForwardPlugin/manifest.json [hunk 1]
```diff
@@ -52,10 +52,20 @@
...
     {
       "key": "signMessage",
       "default": "0",
-      "description": "0 if False and 1 if True indicating whether should sign outgoing J2735 messages."
+      "description": "0 if False and 1 if True, indicating whether Active File will indicate to sign outgoing J2735 messages."
+    },
+    {
+      "key": "EnableHSM",
+      "default": "0",
+      "description": "0 if False and 1 if True, indicating whether HSM is connected and will sign outgoing J2735 messages."
```
Comment (dan-du-car) [inline src/v2i-hub/ImmediateForwardPlugin/manifest.json:68]: Is True HSM connected and sign the messages? What does HSM stand for?

File: src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h [hunk 2]
```diff
@@ -62,10 +62,12 @@ class ImmediateForwardPlugin : public tmx::utils::PluginClient
...
 	std::map<std::string, int> _messageCountMap;
-	std::string signatureData; 
-	std::string url; 
-	std::string baseurl; 
-	unsigned int signState; 
+	std::string signatureData;
+	std::string url;
+	std::string baseurl;
+	std::string txMode;
+	unsigned int signState;
```
Comment (dan-du-car) [inline src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: Does we still need any signState if replaced by enableHSM?

File: src/v2i-hub/ImmediateForwardPlugin/manifest.json (hunk 1 above)
Comment (jwillmartin) [reply src/v2i-hub/ImmediateForwardPlugin/manifest.json:68]: Yeah, true means V2X Hub would try to sign the messages using a Hardware Security Module (only worked with one specific module).

File: src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h (hunk 2 above)
Comment (jwillmartin) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: In most cases, we don't sign the message and an RSU has to sign it. The RSU will expect to receive the Active Message file with this field set to True. So, this allows us to select True without trying to sign the message.

Comment (dan-du-car) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: - Is this Active Message file field set by IMF enableHSM? Or ? 
- Are you saying that we enableHSM = true without sign the message and the signState could be 1?
I am not sure I understand the concept of the enableHSM and signState and if there are any relationships between them. I see this comment before the if (enableHSM=true): `/// if signing is Enabled, request signing with HSM`. It sounds like there are some relationship between them?

Comment (jwillmartin) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: The field in the Active Message file is set by signState.
If you want V2X Hub to sign messages, you have to set enableHSM to 1. You should also set signState to 1 if your RSU is expecting to receive a signed message or will be signing a message. In some cases, you'll want to set both to 1, but in most cases, you'll only want to set signState to 1.

Comment (dan-du-car) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: If we are in most cases and we have enabledHSM to false, signState = 1, is this signState field value 1 updated in the Active Message File? 

Is this code here populate the active message file or?
`
os << "Signature=" << (signState == 1 ? "True" : "False") << "\n" << "Encryption=False\n";
`

Comment (jwillmartin) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: Yep! That's where it's updated

Comment (dan-du-car) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: Thanks! but if enabledHSM = false, this Signature is not populated.

Comment (jwillmartin) [reply src/v2i-hub/ImmediateForwardPlugin/src/ImmediateForwardPlugin.h:69]: The entry for Signature is independent from enabledHSM. It's outside of that if-else case and only relies on signState.

**Review by dan-du-car — APPROVED** [review]

**Pull request merged**